man/syncthing.1
.\" Man page generated from reStructuredText.
.
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.TH "SYNCTHING" "1" "Sep 10, 2026" "v2.1.0" "Syncthing"
.SH NAME
syncthing \- Syncthing
.SH SYNOPSIS
.INDENT 0.0
.INDENT 3.5
.sp
.EX
syncthing [serve]
          [\-\-home=<dir> | \-\-config=<dir> \-\-data=<dir>]
          [\-\-allow\-newer\-config] [\-\-audit] [\-\-auditfile=<file|\-|\-\->]
          [\-\-db\-maintenance\-interval=<interval>]
          [\-\-db\-delete\-retention\-interval=<interval>]
          [\-\-gui\-address=<address>] [\-\-gui\-apikey=<key>]
          [\-\-log\-level=<level>] [\-\-log\-file=<filename>]
          [\-\-log\-max\-old\-files=<num>] [\-\-log\-max\-size=<num>]
          [\-\-log\-format\-timestamp=<format>] [\-\-no\-log\-format\-level\-string]
          [\-\-log\-format\-level\-syslog]
          [\-\-no\-browser] [\-\-no\-console]
          [\-\-no\-port\-probing] [\-\-no\-restart] [\-\-no\-upgrade]
          [\-\-paused] [\-\-unpaused]
          [\-\-version] [\-\-help] [\-\-debug\-*]

syncthing cli
          [\-\-home=<dir> | \-\-config=<dir> \-\-data=<dir>]
          [\-\-gui\-address=<address>] [\-\-gui\-apikey=<key>]
          [\-\-help]
          <command> [command options...] [arguments...]

syncthing browser
          [\-\-home=<dir> | \-\-config=<dir> \-\-data=<dir>]
          [\-\-help]

syncthing decrypt (\-\-to=<dir> | \-\-verify\-only)
          [\-\-home=<dir> | \-\-config=<dir> \-\-data=<dir>]
          [\-\-password=<pw>] [\-\-folder\-id=<id>] [\-\-token\-path=<file>]
          [\-\-continue] [\-\-verbose] [\-\-help]
          <path>

syncthing device\-id
          [\-\-home=<dir> | \-\-config=<dir> \-\-data=<dir>]
          [\-\-help]

syncthing generate
          [\-\-home=<dir> | \-\-config=<dir> \-\-data=<dir>]
          [\-\-gui\-user=<username>] [\-\-gui\-password=<password|\->]
          [\-\-no\-port\-probing]
          [\-\-help]

syncthing paths
          [\-\-home=<dir> | \-\-config=<dir> \-\-data=<dir>]
          [\-\-help]

syncthing upgrade
          [\-\-home=<dir> | \-\-config=<dir> \-\-data=<dir>]
          [\-\-check\-only] [\-\-from=<url>]
          [\-\-help]

syncthing version
          [\-\-home=<dir> | \-\-config=<dir> \-\-data=<dir>]
          [\-\-help]

syncthing debug <command>
          [\-\-home=<dir> | \-\-config=<dir> \-\-data=<dir>]
          [\-\-help]
.EE
.UNINDENT
.UNINDENT
.SH DESCRIPTION
.sp
Syncthing lets you synchronize your files bidirectionally across multiple
devices. This means the creation, modification or deletion of files on one
machine will automatically be replicated to your other devices. We believe your
data is your data alone and you deserve to choose where it is stored. Therefore
Syncthing does not upload your data to the cloud but exchanges your data across
your machines as soon as they are online at the same time.
.sp
The \fBsyncthing\fP core application is a command\-line program which usually runs
in the background and handles the synchronization. It provides a built\-in, HTML
and JavaScript based user interface to be controlled from a web browser. This
frontend communicates with the core application through some HTTP APIs, which
other apps like graphical system integration helpers can use as well, for
greatest flexibility. A link to reach the GUI and API is printed among the first
few log messages.
.SH COMMON OPTIONS
.INDENT 0.0
.TP
.B \-\-home=<dir>, \-H <dir> ($STHOMEDIR)
Set common configuration and data directory. The default configuration
directory is \fB$XDG_STATE_HOME/syncthing\fP or
\fB$HOME/.local/state/syncthing\fP (Unix\-like),
\fB$HOME/Library/Application Support/Syncthing\fP (Mac) and
\fB%LOCALAPPDATA%\eSyncthing\fP (Windows).
.UNINDENT
.INDENT 0.0
.TP
.B \-\-config=<dir>, \-C <dir> ($STCONFDIR)
Set configuration directory. Alternative to \fB\-\-home\fP and must be used
together with \fB\-\-data\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-data=<dir>, \-D <dir> ($STDATADIR)
Set data (e.g. database) directory. Alternative to \fB\-\-home\fP and must be used
together with \fB\-\-config\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-help, \-h
Show help text about command line usage.  Context\-sensitive depending on the
given subcommand.
.UNINDENT
.SH SERVE OPTIONS
.INDENT 0.0
.TP
.B \-\-allow\-newer\-config ($STALLOWNEWERCONFIG)
Try loading a config file written by a newer program version, instead of
failing immediately.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-audit ($STAUDIT)
Write events to timestamped file \fBaudit\-YYYYMMDD\-HHMMSS.log\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-auditfile=<file|\-|\-\-> ($STAUDITFILE)
Use specified file or stream (\fB\(dq\-\(dq\fP for stdout, \fB\(dq\-\-\(dq\fP for stderr) for
audit events, rather than the timestamped default file name.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-db\-maintenance\-interval=<interval> ($STDBMAINTENANCEINTERVAL)
Database maintenance interval – internal database maintenance routines
run this often. The format is that of a Go duration string (see
\X'tty: link https://pkg.go.dev/time#ParseDuration'\fI\%https://pkg.go.dev/time#ParseDuration\fP\X'tty: link').
.UNINDENT
.INDENT 0.0
.TP
.B \-\-db\-delete\-retention\-interval=<interval> ($STDBDELETERETENTIONINTERVAL)
Database deleted item retention interval – deleted items are forgotten
from the database after this interval. The format is that of a Go duration
string (see \X'tty: link https://pkg.go.dev/time#ParseDuration'\fI\%https://pkg.go.dev/time#ParseDuration\fP\X'tty: link').
.UNINDENT
.INDENT 0.0
.TP
.B \-\-gui\-address=<address> ($STGUIADDRESS)
Override GUI listen address. Set this to an address (\fB0.0.0.0:8384\fP)
or a URL (\fBhttp://0.0.0.0:8384\fP). Supported schemes are \fBhttp\fP for
plain HTTP, \fBhttps\fP for HTTP over TLS, \fBunix\fP for plain Unix sockets
or \fBunixs\fP for TLS over Unix sockets. A Unix socket could look like this:
\fBunix:///run/syncthing/syncthing.socket\fP (notice the three slashes: two
as part of the URL structure, one to specify an absolute path).
.UNINDENT
.INDENT 0.0
.TP
.B \-\-gui\-apikey=<string> ($STGUIAPIKEY)
Override the API key needed to access the GUI / REST API.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-log\-file=<filename> ($STLOGFILE)
Set destination filename for logging (use \fB\(dq\-\(dq\fP for stdout, which is the
default option).
.UNINDENT
.INDENT 0.0
.TP
.B \-\-log\-level=<level> ($STLOGLEVEL)
Set the log level for all packages. Valid levels are DEBUG, INFO, WARN,
and ERROR.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-log\-max\-old\-files=<num> ($STLOGMAXOLDFILES)
Number of old files to keep (zero to keep only current).  Applies only when
log rotation is enabled through \fB\-\-log\-max\-size\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-log\-max\-size=<num> ($STLOGMAXSIZE)
Maximum size in bytes of any log file (zero to disable log rotation).
.UNINDENT
.INDENT 0.0
.TP
.B \-\-log\-format\-timestamp=<format> ($STLOGFORMATTIMESTAMP)
Format for timestamp, set to empty to disable timestamps. The format is that
of the Go time formatter (see \X'tty: link https://pkg.go.dev/time#Layout'\fI\%https://pkg.go.dev/time#Layout\fP\X'tty: link').
.UNINDENT
.INDENT 0.0
.TP
.B \-\-[no\-]log\-format\-level\-string ($STLOGFORMATLEVELSTRING)
Whether to include level string (e.g. “INF”) in log line.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-[no\-]log\-format\-level\-syslog ($STLOGFORMATLEVELSYSLOG)
Whether to include level as syslog prefix (e.g. “<6>”) in log line.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-no\-browser ($STNOBROWSER)
Do not start a browser.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-no\-console ($STNOCONSOLE)
Hide the console window. (On Windows only)
.UNINDENT
.INDENT 0.0
.TP
.B \-\-no\-port\-probing ($STNOPORTPROBING)
Don’t try to find unused random ports for the GUI and listen address when
generating an initial configuration / starting for the first time.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-no\-restart ($STNORESTART)
Do not restart Syncthing when it exits. The monitor process will still run
to handle crashes and writing to logfiles (if configured to).
.UNINDENT
.INDENT 0.0
.TP
.B \-\-no\-upgrade ($STNOUPGRADE)
Disable automatic upgrades.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-paused ($STPAUSED)
Start with all devices and folders paused.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-unpaused ($STUNPAUSED)
Start with all devices and folders unpaused.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-version
Show the current version information, then exit.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-debug\-*
Several debug options exist. See \fBsyncthing serve \-\-help\fP for the flags
that apply to your version.
.UNINDENT
.SH DECRYPT OPTIONS
.INDENT 0.0
.TP
.B \-\-to=<dir>
Destination directory where files should be stored after decryption.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-verify\-only
Don’t write decrypted files to disk (but verify plaintext hashes).
.UNINDENT
.INDENT 0.0
.TP
.B \-\-password=<pw> ($FOLDER_PASSWORD)
Folder password for decryption / verification.  Can be passed through the
\fBFOLDER_PASSWORD\fP environment variable instead to avoid recording in a
shell’s history buffer or sniffing from the running processes list.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-folder\-id=<id>
Folder ID of the encrypted folder, if it cannot be determined automatically.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-token\-path=<file>
Path to the token file within the folder (used to determine folder ID).
.UNINDENT
.INDENT 0.0
.TP
.B \-\-continue
Continue processing next file in case of error, instead of aborting.
.UNINDENT
.SH GENERATE OPTIONS
.INDENT 0.0
.TP
.B \-\-gui\-password=<password|\->
Specify new GUI authentication password, to update the config file.  Read
from the standard input stream if only a single dash (\fB\-\fP) is given.  A
plaintext password is hashed before writing to the config file, but an
already bcrypt\-hashed input is stored verbatim.  As a special case, giving
the existing password hash as password will leave it untouched.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-gui\-user=<username>
Specify new GUI authentication user name, to update the config file.
.UNINDENT
.SH UPGRADE OPTIONS
.INDENT 0.0
.TP
.B \-\-check\-only
Do not upgrade, only indicate whether an upgrade is available.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-from=<url>
Upgrade to the Syncthing version available from loading the package at
the given URL.
.UNINDENT
.SH EXIT CODES
.INDENT 0.0
.TP
.B 0
Success / Shutdown
.TP
.B 1
Error
.TP
.B 2
Upgrade not available
.TP
.B 3
Restarting
.TP
.B 4
Upgrading
.UNINDENT
.sp
Exit codes over 125 are usually returned by the shell/binary loader/default
signal handler. Exit codes over 128+N on Unix usually represent the signal which
caused the process to exit. For example, \fB128 + 9 (SIGKILL) = 137\fP\&.
.SH SUBCOMMANDS
.sp
The command line syntax supports different modes of operation through
several subcommands, specified as the first argument.  If omitted, the
default \fBserve\fP is assumed.
.sp
The initial setup of a device ID and default configuration can be called
explicitly with the \fBgenerate\fP subcommand.  It can also update the configured
GUI authentication credentials, without going through the REST API.  An existing
device certificate is left untouched.  If the configuration file already exists,
it is validated and updated to the latest configuration schema, including adding
default values for any new options.
.sp
The \fBdecrypt\fP subcommand is used in conjunction with untrusted (encrypted)
devices, see the relevant section on \fI\%decryption\fP for
details.  It does not depend on Syncthing to be running, but works on offline
data.
.sp
To work with the REST API for debugging or automating things in Syncthing, the
\fBcli\fP subcommand provides easy access to individual features.  It basically
saves the hassle of handling HTTP connections and API authentication.
.sp
The available subcommands are grouped into several nested hierarchies and some
parts dynamically generated from the running Syncthing instance.  On every
level, the \fB\-\-help\fP option lists the available properties, actions and
commands for the user to discover interactively.  The top\-level groups are:
.INDENT 0.0
.TP
.B config
Access the live configuration in a running instance over the REST API to
retrieve (get) or update (set) values in a fine\-grained way.  The hierarchy
is based on the same structure as used in the JSON / XML representations.
.TP
.B show
Show system properties and status of a running instance.  The output is
passed on directly from the REST API response and therefore requires parsing
JSON format.
.TP
.B operations
Control the overall program operation such as restarting or handling
upgrades, as well as triggering some actions on a per\-folder basis.  Can
also configure the default ignore patterns from a plain text ignore file.
.TP
.B errors
Examine pending error conditions that need attention from the user, or
acknowledge (clear) them.
.TP
.B debug
Various tools to aid in diagnosing problems or collection information for
bug reports.
.TP
.B \fB\-\fP (a single dash)
Reads subsequent commands from the standard input stream, without needing to
call the \fBsyncthing cli\fP command over and over.  Exits on any invalid
command or when EOF (end\-of\-file) is received.
.UNINDENT
.sp
For troubleshooting issues, the \fBdebug\fP subcommand provides some commonly
needed actions for data analysis and repair.  Some of these commands access the
database directly and can therefore only work when Syncthing is not running.
Not to be confused with the \fBcli debug\fP subcommand group, which works via the
REST API.
.INDENT 0.0
.TP
.B debug reset\-database
Reset the database, forcing a full rescan and resync.  \fBMust only be used
when Syncthing is not running.\fP  More granular reset operations are
available on the REST API while Syncthing is running:
\fI\%POST /rest/system/reset\fP
.TP
.B debug database\-statistics
Display database size statistics.
.TP
.B debug database\-counts
Display database folder counts.
.TP
.B debug database\-file
Display database file metadata.
.UNINDENT
.SH PROXIES
.sp
Syncthing can use a SOCKS, HTTP, or HTTPS proxy to talk to the outside
world. The proxy is used for outgoing connections only \- it is not possible
to accept incoming connections through the proxy. The proxy is configured
through the environment variable \fBALL_PROXY\fP or its lowercase variant
\fBall_proxy\fP\&. If both are set, \fBALL_PROXY\fP takes precedence. For example:
.INDENT 0.0
.INDENT 3.5
.sp
.EX
$ export ALL_PROXY=socks://192.0.2.42:8081
.EE
.UNINDENT
.UNINDENT
.SH DEVELOPMENT SETTINGS
.sp
The following environment variables modify Syncthing’s behavior in ways that
are mostly useful for developers. Use with care.
If you start Syncthing from within service managers like systemd or supervisor,
path expansion may not be supported.
.INDENT 0.0
.TP
.B STTRACE
Used to increase the debugging verbosity in specific facilities,
generally mapping to a Go package. Enter a comma\-separated string of
facilities to trace: \fBapi,beacon\fP\&. Optionally, a log level can be
given per facility to specify something other than DEBUG:
\fBapi:WARN,beacon:ERROR\fP, potentially overriding a global \fB\-\-log\-level\fP
adjustment.
.sp
The valid facility strings are listed below; additionally, \fBsyncthing
serve \-\-help\fP always outputs the most up\-to\-date list.
.INDENT 7.0
.TP
.B api
REST API
.TP
.B beacon
Multicast and broadcast discovery
.TP
.B config
Configuration loading and saving
.TP
.B connections
Connection handling
.TP
.B db/sqlite
SQLite database
.TP
.B dialer
Dialing connections
.TP
.B discover
Remote device discovery
.TP
.B events
Event generation and logging
.TP
.B fs
Filesystem access
.TP
.B main
Main package
.TP
.B model
The root hub
.TP
.B nat
NAT discovery and port mapping
.TP
.B pmp
NAT\-PMP discovery and port mapping
.TP
.B protocol
The BEP protocol
.TP
.B relay/client
Relay client
.TP
.B scanner
File change detection and hashing
.TP
.B stun
STUN functionality
.TP
.B syncthing
Main run facility
.TP
.B upgrade
Binary upgrades
.TP
.B upnp
UPnP discovery and port mapping
.TP
.B ur
Usage reporting
.TP
.B versioner
File versioning
.TP
.B watchaggregator
Filesystem event watcher
.UNINDENT
.TP
.B STLOCKTHRESHOLD
Used for debugging internal deadlocks; sets debug sensitivity. Use only
under direction of a developer.
.TP

debian/changelog
syncthing ({{.version}}); urgency=medium

  * Packaging of {{.version}}.

 -- Syncthing Release Management <release@syncthing.net>  {{.date}}


lib/syncthing/syncthing.go
// Copyright (C) 2014 The Syncthing Authors.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this file,
// You can obtain one at https://mozilla.org/MPL/2.0/.

package syncthing

import (
	"context"
	"crypto/tls"
	"errors"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"os"
	"runtime"
	"slices"
	"strings"
	"sync"
	"time"

	"github.com/thejerf/suture/v4"

	"github.com/syncthing/syncthing/internal/db"
	"github.com/syncthing/syncthing/internal/slogutil"
	"github.com/syncthing/syncthing/lib/api"
	"github.com/syncthing/syncthing/lib/build"
	"github.com/syncthing/syncthing/lib/config"
	"github.com/syncthing/syncthing/lib/connections"
	"github.com/syncthing/syncthing/lib/connections/registry"
	"github.com/syncthing/syncthing/lib/discover"
	"github.com/syncthing/syncthing/lib/events"
	"github.com/syncthing/syncthing/lib/locations"
	"github.com/syncthing/syncthing/lib/model"
	"github.com/syncthing/syncthing/lib/osutil"
	"github.com/syncthing/syncthing/lib/protocol"
	"github.com/syncthing/syncthing/lib/svcutil"
	"github.com/syncthing/syncthing/lib/tlsutil"
	"github.com/syncthing/syncthing/lib/upgrade"
	"github.com/syncthing/syncthing/lib/ur"
)

const (
	bepProtocolName        = "bep/1.0"
	tlsDefaultCommonName   = "syncthing"
	maxSystemErrors        = 5
	initialSystemLog       = 10
	maxSystemLog           = 250
	deviceCertLifetimeDays = 20 * 365
)

type Options struct {
	AuditWriter           io.Writer
	NoUpgrade             bool
	ProfilerAddr          string
	ResetDeltaIdxs        bool
	DBMaintenanceInterval time.Duration
}

type App struct {
	myID              protocol.DeviceID
	mainService       *suture.Supervisor
	cfg               config.Wrapper
	sdb               db.DB
	evLogger          events.Logger
	cert              tls.Certificate
	opts              Options
	exitStatus        svcutil.ExitStatus
	err               error
	stopOnce          sync.Once
	mainServiceCancel context.CancelFunc
	stopped           chan struct{}
	dbService         db.DBService

	// Access to internals for direct users of this package. Note that the interface in Internals is unstable!
	Internals *Internals
}

func New(cfg config.Wrapper, sdb db.DB, evLogger events.Logger, cert tls.Certificate, opts Options) (*App, error) {
	a := &App{
		cfg:      cfg,
		sdb:      sdb,
		evLogger: evLogger,
		opts:     opts,
		cert:     cert,
		stopped:  make(chan struct{}),
	}
	close(a.stopped) // Hasn't been started, so shouldn't block on Wait.
	return a, nil
}

// Start executes the app and returns once all the startup operations are done,
// e.g. the API is ready for use.
// Must be called once only.
func (a *App) Start() error {
	// Create a main service manager. We'll add things to this as we go along.
	// We want any logging it does to go through our log system.
	spec := svcutil.SpecWithDebugLogger()
	a.mainService = suture.New("main", spec)

	// Start the supervisor and wait for it to stop to handle cleanup.
	a.stopped = make(chan struct{})
	ctx, cancel := context.WithCancel(context.Background())
	a.mainServiceCancel = cancel
	errChan := a.mainService.ServeBackground(ctx)
	go a.wait(errChan)

	if err := a.startup(); err != nil {
		a.stopWithErr(svcutil.ExitError, err)
		return err
	}

	return nil
}

// StartMaintenance asynchronously triggers database maintenance to start.
func (a *App) StartMaintenance() <-chan error {
	return a.dbService.StartMaintenance()
}

// LastMaintenanceTime returns the last time database maintenance completed successfully
// This will return time zero when database maintenance has never completed successfully.
func (a *App) LastMaintenanceTime() time.Time {
	return a.dbService.LastMaintenanceTime()
}

func (a *App) startup() error {
	a.mainService.Add(ur.NewFailureHandler(a.cfg, a.evLogger))

	a.dbService = a.sdb.Service(a.opts.DBMaintenanceInterval)
	a.mainService.Add(a.dbService)

	if a.opts.AuditWriter != nil {
		a.mainService.Add(newAuditService(a.opts.AuditWriter, a.evLogger))
	}

	// Event subscription for the API; must start early to catch the early
	// events. The LocalChangeDetected event might overwhelm the event
	// receiver in some situations so we will not subscribe to it here.
	defaultSub := events.NewBufferedSubscription(a.evLogger.Subscribe(api.DefaultEventMask), api.EventSubBufferSize)
	diskSub := events.NewBufferedSubscription(a.evLogger.Subscribe(api.DiskEventMask), api.EventSubBufferSize)

	// Attempt to increase the limit on number of open files to the maximum
	// allowed, in case we have many peers. We don't really care enough to
	// report the error if there is one.
	osutil.MaximizeOpenFileLimit()

	// Figure out our device ID and log it.
	a.myID = protocol.NewDeviceID(a.cert.Certificate[0])
	slog.Info("Calculated our device ID", slog.String("device", a.myID.String()))

	// Emit the Starting event, now that we know who we are.

	a.evLogger.Log(events.Starting, map[string]string{
		"home": locations.GetBaseDir(locations.ConfigBaseDir),
		"myID": a.myID.String(),
	})

	if err := checkShortIDs(a.cfg); err != nil {
		slog.Error("Short device IDs are in conflict; regenerate the device ID of one of the conflicting devices", slogutil.Error(err))
		return err
	}

	if len(a.opts.ProfilerAddr) > 0 {
		go func() {
			l.Debugln("Starting profiler on", a.opts.ProfilerAddr)
			runtime.SetBlockProfileRate(1)
			err := http.ListenAndServe(a.opts.ProfilerAddr, nil)
			if err != nil {
				slog.Warn("Failed to listen and serve for profiles", slogutil.Error(err))
				return
			}
		}()
	}

	if slog.Default().Enabled(context.Background(), slog.LevelInfo) {
		go func() {
			perf := ur.CpuBench(context.Background(), 3, 150*time.Millisecond)
			slog.Info("Measured hashing performance", "perf", fmt.Sprintf("%.02f MB/s", perf))
		}()
	}

	if a.opts.ResetDeltaIdxs {
		slog.Info("Reinitializing delta index IDs")
		if err := a.sdb.DropAllIndexIDs(); err != nil {
			slog.Error("Failed to drop index IDs", slogutil.Error(err))
			return err
		}
	}

	protectedFiles := []string{
		locations.Get(locations.Database),
		locations.Get(locations.ConfigFile),
		locations.Get(locations.CertFile),
		locations.Get(locations.KeyFile),
	}

	// Remove database entries for folders that no longer exist in the config
	cfgFolders := a.cfg.Folders()
	dbFolders, err := a.sdb.ListFolders()
	if err != nil {
		slog.Warn("Failed to list folders", slogutil.Error(err))
		return err
	}
	for _, folder := range dbFolders {
		if _, ok := cfgFolders[folder]; !ok {
			slog.Info("Cleaning metadata for dropped folder", "folder", folder)
			a.sdb.DropFolder(folder)
		}
	}

	// Grab the previously running version string from the database.

	miscDB := db.NewMiscDB(a.sdb)
	prevVersion, _, err := miscDB.String("prevVersion")
	if err != nil {
		slog.Error("Database error when getting previous version", slogutil.Error(err))
		return err
	}

	// Strip away prerelease/beta stuff and just compare the release
	// numbers. 0.14.44 to 0.14.45-banana is an upgrade, 0.14.45-banana to
	// 0.14.45-pineapple is not.

	prevParts := strings.Split(prevVersion, "-")
	curParts := strings.Split(build.Version, "-")
	if rel := upgrade.CompareVersions(prevParts[0], curParts[0]); rel != upgrade.Equal {
		if prevVersion != "" {
			slog.Info("Detected upgrade", "from", prevVersion, "to", build.Version)
		}

		if a.cfg.Options().SendFullIndexOnUpgrade {
			// Drop delta indexes in case we've changed random stuff we
			// shouldn't have. We will resend our index on next connect.
			if err := a.sdb.DropAllIndexIDs(); err != nil {
				slog.Warn("Failed to drop index IDs", slogutil.Error(err))
				return err
			}
		}
	}

	if build.Version != prevVersion {
		// Remember the new version.
		miscDB.PutString("prevVersion", build.Version)
	}

	if err := globalMigration(a.sdb, a.cfg); err != nil {
		slog.Warn("Failed to perform global migration", slogutil.Error(err))
		return err
	}

	keyGen := protocol.NewKeyGenerator()
	m := model.NewModel(a.cfg, a.myID, a.sdb, protectedFiles, a.evLogger, keyGen)
	a.Internals = newInternals(m)

	a.mainService.Add(m)

	// The TLS configuration is used for both the listening socket and outgoing
	// connections.

	tlsCfg := tlsutil.SecureDefaultTLS13()
	tlsCfg.Certificates = []tls.Certificate{a.cert}
	tlsCfg.NextProtos = []string{bepProtocolName}
	tlsCfg.ClientAuth = tls.RequestClientCert
	tlsCfg.SessionTicketsDisabled = true
	tlsCfg.InsecureSkipVerify = true

	// Start discovery and connection management

	// Chicken and egg, discovery manager depends on connection service to tell it what addresses it's listening on
	// Connection service depends on discovery manager to get addresses to connect to.
	// Create a wrapper that is then wired after they are both set up.
	addrLister := &lateAddressLister{}

	connRegistry := registry.New()
	discoveryManager := discover.NewManager(a.myID, a.cfg, a.cert, a.evLogger, addrLister, connRegistry)
	connectionsService := connections.NewService(a.cfg, a.myID, m, tlsCfg, discoveryManager, bepProtocolName, tlsDefaultCommonName, a.evLogger, connRegistry, keyGen)

	addrLister.AddressLister = connectionsService

	a.mainService.Add(discoveryManager)
	a.mainService.Add(connectionsService)

	a.cfg.Modify(func(cfg *config.Configuration) {
		// Candidate builds always run with usage reporting.
		if build.IsCandidate {
			slog.Info("Anonymous usage reporting is always enabled for candidate releases")
			if cfg.Options.URAccepted != ur.Version {
				cfg.Options.URAccepted = ur.Version
				// Unique ID will be set and config saved below if necessary.
			}
		}
	})

	usageReportingSvc := ur.New(a.cfg, m, connectionsService, a.opts.NoUpgrade)
	a.mainService.Add(usageReportingSvc)

	// GUI

	if err := a.setupGUI(m, defaultSub, diskSub, discoveryManager, connectionsService, usageReportingSvc, slogutil.ErrorRecorder, slogutil.GlobalRecorder, miscDB); err != nil {
		slog.Error("Failed to start API", slogutil.Error(err))
		return err
	}

	myDev, _ := a.cfg.Device(a.myID)
	slog.Info("Loaded configuration", "name", myDev.Name)
	for _, device := range a.cfg.Devices() {
		if device.DeviceID != a.myID {
			slog.Info("Loaded peer device configuration", device.DeviceID.LogAttr(), slog.String("name", device.Name), slogutil.Address(device.Addresses))
		}
	}

	if isSuperUser() {
		slog.Warn("Syncthing should not run as a privileged or system user; please consider using a normal user account")
	}

	a.evLogger.Log(events.StartupComplete, map[string]string{
		"myID": a.myID.String(),
	})

	if a.cfg.Options().SetLowPriority {
		if err := osutil.SetLowPriority(); err != nil {
			slog.Warn("Failed to lower process priority", slogutil.Error(err))
		}
	}

	return nil
}

func (a *App) wait(errChan <-chan error) {
	err := <-errChan
	a.handleMainServiceError(err)

	done := make(chan struct{})
	go func() {
		a.sdb.Close()
		close(done)
	}()
	select {
	case <-done:
	case <-time.After(10 * time.Second):
		slog.Warn("Database failed to stop within 10s")
	}

	slog.Info("Exiting")

	close(a.stopped)
}

func (a *App) handleMainServiceError(err error) {
	if err == nil || errors.Is(err, context.Canceled) {
		return
	}
	var fatalErr *svcutil.FatalErr
	if errors.As(err, &fatalErr) {
		a.exitStatus = fatalErr.Status
		a.err = fatalErr.Err
		return
	}
	a.err = err
	a.exitStatus = svcutil.ExitError
}

// Wait blocks until the app stops running. Also returns if the app hasn't been
// started yet.
func (a *App) Wait() svcutil.ExitStatus {
	<-a.stopped
	return a.exitStatus
}

// Error returns an error if one occurred while running the app. It does not wait
// for the app to stop before returning.
func (a *App) Error() error {
	select {
	case <-a.stopped:
		return a.err
	default:
	}
	return nil
}

// Stop stops the app and sets its exit status to given reason, unless the app
// was already stopped before. In any case it returns the effective exit status.
func (a *App) Stop(stopReason svcutil.ExitStatus) svcutil.ExitStatus {
	return a.stopWithErr(stopReason, nil)
}

func (a *App) stopWithErr(stopReason svcutil.ExitStatus, err error) svcutil.ExitStatus {
	a.stopOnce.Do(func() {
		a.exitStatus = stopReason
		a.err = err
		if shouldDebug() {
			slog.Debug("Services before stop:")
			printServiceTree(os.Stdout, a.mainService, 0)
		}
		a.mainServiceCancel()
	})
	<-a.stopped
	return a.exitStatus
}

func (a *App) setupGUI(m model.Model, defaultSub, diskSub events.BufferedSubscription, discoverer discover.Manager, connectionsService connections.Service, urService *ur.Service, errors, systemLog slogutil.Recorder, miscDB *db.Typed) error {
	guiCfg := a.cfg.GUI()

	if !guiCfg.Enabled {
		return nil
	}

	if guiCfg.InsecureAdminAccess {
		slog.Warn("Insecure admin access is enabled")
	}

	summaryService := model.NewFolderSummaryService(a.cfg, m, a.myID, a.evLogger)
	a.mainService.Add(summaryService)

	apiSvc := api.New(a.myID, a.cfg, locations.Get(locations.GUIAssets), tlsDefaultCommonName, m, defaultSub, diskSub, a.evLogger, discoverer, connectionsService, urService, summaryService, errors, systemLog, a.opts.NoUpgrade, miscDB)
	a.mainService.Add(apiSvc)

	if err := apiSvc.WaitForStart(); err != nil {
		return err
	}
	return nil
}

// checkShortIDs verifies that the configuration won't result in duplicate
// short ID:s; that is, that the devices in the cluster all have unique
// initial 64 bits.
func checkShortIDs(cfg config.Wrapper) error {
	exists := make(map[protocol.ShortID]protocol.DeviceID)
	for deviceID := range cfg.Devices() {
		shortID := deviceID.Short()
		if otherID, ok := exists[shortID]; ok {
			return fmt.Errorf("%v in conflict with %v", deviceID, otherID)
		}
		exists[shortID] = deviceID
	}
	return nil
}

type supervisor interface{ Services() []suture.Service }

func printServiceTree(w io.Writer, sup supervisor, level int) {
	printService(w, sup, level)

	svcs := sup.Services()
	slices.SortFunc(svcs, func(a, b suture.Service) int {
		return strings.Compare(fmt.Sprint(a), fmt.Sprint(b))
	})

	for _, svc := range svcs {
		if sub, ok := svc.(supervisor); ok {
			printServiceTree(w, sub, level+1)
		} else {
			printService(w, svc, level+1)
		}
	}
}

func printService(w io.Writer, svc any, level int) {
	type errorer interface{ Error() error }

	t := "-"
	if _, ok := svc.(supervisor); ok {
		t = "+"
	}
	fmt.Fprintln(w, strings.Repeat("  ", level), t, svc)
	if es, ok := svc.(errorer); ok {
		if err := es.Error(); err != nil {
			fmt.Fprintln(w, strings.Repeat("  ", level), "  ->", err)
		}
	}
}

type lateAddressLister struct {
	discover.AddressLister
}
